### PR TITLE
fix: [CDS-86427]: A new internal Api to be used for template api

### PR DESCRIPTION
**Pull request by abhinavmittal-harness**
## Describe your changes

As part of the api, we are basically clubbing functionality for two different api.
Get Template and Apply template.

The reason for this is for instance we create a shell template with shell type as PowerShell.
In the same template we have a field executionTarget.connectorRef as a runtimeInput.
This field is dependent on Shell Type. For bash it is gonna be ssh credentials and for powershell it is gonna be winrm credentials.

Now when the step template is used in stageTemplate or pipeline Template, the yaml doesn't contain any information of the shell type. It only contains template Input which would only have executionTarget.connectorRef. There by ui not able to figure out what display name should be shown for the connector Ref.

To solve this we are introducing a new api to get resolved template from get go. And the reason we haven't modified get Template api as, that is exposed via api docs and terraform and we want to keep this call hidden from those resources.

Fetched getTemplates with templatesResolvedYaml as true and false locally
Getting The response being populated with mergedTemplateYaml and null respecitvely

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)

### CODE REVIEW

**Comment by abhinavmittal-harness** [comment]
trigger uts

File: template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java [hunk 1]
```diff
@@ -704,4 +706,34 @@ ResponseDTO<PageResponse<EntitySetupUsageDTO>> listTemplateEntityUsage(
           NGCommonEntityConstants.VERSION_LABEL_KEY) String versionLabel,
       @Parameter(description = "This contains details of Git Entity like Git Branch info to be updated")
       TemplateUpdateGitMetadataRequest request);
+
+  @POST
+  @Path("/internal/{templateIdentifier}")
+  @ApiOperation(value = "Gets a merged template by identifier", nickname = "Get Resolved Template")
```
Comment (abhinavmittal-harness) [inline template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:712]: Gets a resolved template by identifier

File: template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java [hunk 2]
```diff
@@ -704,4 +706,34 @@ ResponseDTO<PageResponse<EntitySetupUsageDTO>> listTemplateEntityUsage(
...
+  @Operation(operationId = "getResolvedTemplate", description = "Returns a Template by Identifier based on userInput",
```
Comment (abhinavmittal-harness) [inline template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:713]: Returns the merged template by identifier

File: template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java [hunk 3]
```diff
@@ -704,4 +706,34 @@ ResponseDTO<PageResponse<EntitySetupUsageDTO>> listTemplateEntityUsage(
...
+      summary = "Fetch a Resolved Template",
```
Comment (abhinavmittal-harness) [inline template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:714]: Fetch merged template by identifier

File: template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java [hunk 4]
```diff
@@ -704,4 +706,34 @@ ResponseDTO<PageResponse<EntitySetupUsageDTO>> listTemplateEntityUsage(
...
+      responses =
+      {
+        @io.swagger.v3.oas.annotations.responses.
+        ApiResponse(responseCode = "default", description = "Returns pipeline YAML")
```
Comment (abhinavmittal-harness) [inline template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:718]: Returns Template DTO with original yaml and resolved Yaml

File: template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mohitg0795) [reply template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java:712]: If its fetching merged template, then why not name the path as well accordingly. Currently from the path, it looks like just getting a template.

**Comment by abhinavmittal-harness** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]

**Comment by sahilhindwani-harness** [comment]
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 724 at r3](https://reviewable.io/reviews/harness/harness-core/56665#-NlXPeLS5dgqeKIaXwNE:-NlXPeLS5dgqeKIaXwNF:bdgrcbh) ([raw file](https://github.com/harness/harness-core/blob/4570a89ad269eecdbe526a9290e245560d7bc6da/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L724)):*
> ```Java
>   @Hidden
>   ResponseDTO<TemplateResponseDTO>
>   getTemplateByIdentifier(@Parameter(description = PipelineResourceConstants.ACCOUNT_PARAM_MESSAGE, required = true)
> ```

nit: can we rename this to something like getResolvedTemplate?

**Review by sahilhindwani-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 5 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by utkarshchoubey-harness — APPROVED** [review]
lgtm!

**Comment by abhinavmittal-harness** [comment]
trigger messagecheck

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 5 file(s) to review. 2 change requests. 1 of 2 approvals obtained. 

___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 711 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZ_U2I-RCFbMXYv_Ao:-NlZ_U2I-RCFbMXYv_Ap:b-qjuebr) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L711)):*
> ```Java
> 
>   @POST
>   @Path("/getResolvedTemplate/{templateIdentifier}")
> ```

path naming should be kebab case

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 5 file(s) to review. 2 change requests. 1 of 2 approvals obtained. 

___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 722 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZ_oFt0FLPKN8pGEA2:-NlZ_oFt0FLPKN8pGEA3:b-iu277z) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L722)):*
> ```Java
>   @Timed
>   @ResponseMetered
>   @Hidden
> ```

If its hidden, then lets not put any swagger annotations for easier readability. No swagger annotation makes sense for hidden API.

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 5 file(s) to review. 2 change requests. 1 of 2 approvals obtained. 

___
*[`template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java` line 578 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZa93-2i0vFjoXPn3B:-NlZa93-2i0vFjoXPn3C:b-sa79x0) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java#L578)):*
> ```Java
>       String templateIdentifier, String versionLabel, boolean deleted, GitEntityFindInfoDTO gitEntityBasicInfo,
>       MergeTemplateRequestDTO mergeTemplateRequestDTO, boolean loadFromFallbackBranch, String loadFromCache) {
>     ResponseDTO<TemplateResponseDTO> templateResponseDTO = get(accountId, orgId, projectId, templateIdentifier,
> ```

lets not put any functional logic in resource layer. All functional logic should go to service layer only.

**Review by abhinavmittal-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 5 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 711 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZ_U2I-RCFbMXYv_Ao:-NlaeKec5Tgomx0ymcxI:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L711)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

path naming should be kebab case
</blockquote></details>

Done.
___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 722 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZ_oFt0FLPKN8pGEA2:-NlaeL4o59U6j_tYz_4y:b-d1eih4) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L722)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

If its hidden, then lets not put any swagger annotations for easier readability. No swagger annotation makes sense for hidden API.
</blockquote></details>

removed swagger 2 annotation
___
*[`template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java` line 578 at r4](https://reviewable.io/reviews/harness/harness-core/56665#-NlZa93-2i0vFjoXPn3B:-NlaeNoGEdC7vg9TyqhP:b-uu1cjo) ([raw file](https://github.com/harness/harness-core/blob/a1d28752643a9ce4c1b9a9f9ccb9dc82004753e4/template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java#L578)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

lets not put any functional logic in resource layer. All functional logic should go to service layer only.
</blockquote></details>

UNfortunately the methods we are using are present int he resource layer.
For me to move the buisness logic to service layer would require larger refactoring.
And considering this is a time sensitive CFD for now  I would prefer keeping it this way..
We can create a tech debt to handle this refactoring later.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 2 of 5 files at r1, 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665)* status: 3 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Review by sahilhindwani-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56665#-:-Nlc9WPo7RGM1oVfJSF3:bnfp4nl)* status: <span>:shipit:</span> complete! ✓

**Comment by abhinavmittal-harness** [comment]
trigger review

**Comment by abhinavmittal-harness** [comment]
trigger uts

**Comment by abhinavmittal-harness** [comment]
trigger ut6,ut7,ut8,ut9

**Comment by abhinavmittal-harness** [comment]
trigger ut6,ut7,ut8

**Comment by abhinavmittal-harness** [comment]
trigger ut6

**Pull request merged**